### PR TITLE
Add Bazel config for zlib support

### PR DESCRIPTION
**Pull request by schroederc**
Allow downstream Bazel clients to build protobuf library with zlib support.

### CODE REVIEW

**Comment by acozzette** [comment]
@schroederc I tried this out locally but got a build error. When I do `bazel build --define protobuf_zlib=true //:protobuf` I get an error saying `no such target '//external:zlib': target 'zlib' not declared in package 'external'` Is that error to be expected?

**Comment by schroederc** [comment]
It's expected on the protobuf side because it does not bind the `zlib` target in its `WORKSPACE`.  It works if you add zlib as a dependency in your WORKSPACE (https://github.com/kythe/kythe/blob/master/external.bzl#L18) and then bind that target (https://github.com/kythe/kythe/blob/master/external.bzl#L606).

It seemed like the protobuf build doesn't care about zlib support so I didn't add it, but downstream builds can enable it if they have zlib.

**Comment by acozzette** [comment]
Ah OK, thanks for the explanation. Is there any way we can just unconditionally depend on `zlib`? I think if possible it would be nice to avoid the need for a new build flag here.

**Comment by schroederc** [comment]
Yes, you'd need to add the zlib repo to the `WORKSPACE` file as mentioned above.  This will have the side-effect of requiring all downstream builds to also add that repository once they update protobuf.  I can do that, if that's wanted instead.

**Comment by acozzette** [comment]
OK, I see. That's too bad that adding the dependency unconditionally would break downstream builds. Is there a best practice in Bazel for how to add a new dependency to a library without disrupting downstream projects? Your approach of using a flag should work, but still it would be nice if we could avoid adding a new flag.

**Comment by schroederc** [comment]
I don't know of a way to add the dependency to all downstream projects automagically.  Many projects use a custom [Starlark function](https://github.com/bazelbuild/rules_go/blob/master/go/workspace.rst#go-rules-dependencies) that downstream can call to add necessary deps to their WORKSPACE.  The upstream project can then curate the list of deps without breaking builds.  However, current downstream projects using protobuf do not already call such a function.

**Comment by acozzette** [comment]
@schroederc OK, thanks for the info. I'd say for now let's go with omitting the flag and making the dependency unconditional. Later on we can add a Starlark function like the one you described if we need to.

File: BUILD [hunk 1]
```diff
@@ -15,6 +15,17 @@ config_setting(
     },
 )
 
+################################################################################
+# ZLIB configuration
+################################################################################
+
+config_setting(
```
Comment (acozzette) [inline BUILD]: We no longer need this new config_setting, right?

Comment (schroederc) [reply BUILD]: Right.

**Comment by acozzette** [comment]
Thanks @schroederc!

**Comment by ronshapiro** [comment]
Is `@bazel_tools//third_party/zlib` an option here? That way there's no additional work for users

**Comment by schroederc** [comment]
Any Bazel repository depending on protobuf could put this in their WORKSPACE to have protobuf compile against the Bazel version of zlib:

```
bind(
    name = "zlib",
    actual = "@bazel_tools//third_party/zlib",
)
```

It's a different question to pin all builds of protobuf to the Bazel version of zlib (which is not under control of protobuf).

**Pull request merged**